Write getting started guide

Lewiky opened the pull request:
Adds the getting started guide, meant to be shallowish walkthrough of the whole process of running an application, breaking down what everything means

jaylees14 commented on TempleEight/docs/getting-started.md:
````diff
@@ -4,3 +4,400 @@ title: Getting Started
+Templefile validated correctly
+```
+
+Which means our Templefile is syntactically correct, and we an begin to generate all the code we need.
````
Maybe don't start a sentence with which

jaylees14 replied:
```suggestion
Which means our Templefile is syntactically correct, and we can begin to generate all the code we need.
```

jaylees14 commented on TempleEight/docs/getting-started.md:
````diff
@@ -4,3 +4,400 @@ title: Getting Started
+
+### Generating Code
+
+At this point, we can begin to generate some code. Running: 
+
+```
+~/Documents/temple-tutorial ❯❯❯ temple generate dating-app.temple
+```
+
+Begins the process. The first thing we'll see is the CLI prompting us to give some further generation details that are specific to the metadata selections made in our Templefile. 
````
I'm also not really a fan of the splitting sentences and then capitalising. 

How about

Running the following command will generate the code for the project:

jaylees14 commented on TempleEight/docs/getting-started.md:
````diff
@@ -4,3 +4,400 @@ title: Getting Started
+What should the Go module name be? (expected format "github.com/username/repo")
+```
+
+We need to enter the name of the Golang module we'll be generating into for our project, this can be anything, but standards dictate it should be the github URL of where this project will be hosted. 
````
```suggestion
We need to enter the name of the Golang module we'll be generating into for our project, this can be anything, but standards dictate it should be the GitHub URL of where this project will be hosted. 
```

jaylees14 commented:
Actually I disagree with myself - I think all this content is definitely needed, although should be broken down - but the "getting started" bit should be much much simpler. So maybe 1 service, with a few fields, nothing special at all. Then introduce the other stuff in later sections? @Lewiky

xsanda commented on TempleEight/docs/getting-started.md:
```diff
@@ -4,3 +4,400 @@ title: Getting Started
+This tutorial will walk you through developing your own applications with Temple, from start to finish.
+
+We're going to assume you've already installed the Temple CLI, as per the [Installation Guide](installation.md). 
+> If you'd like to verify this, run `temple --version` 
```
How do I know if this worked?

xsanda commented on TempleEight/docs/getting-started.md:
```diff
@@ -4,3 +4,400 @@ title: Getting Started
+
+Service block can contain two kinds of declarations: *Property definitions* and *metadata*. 
+
+Property declarations tell us about the kind of data belonging to the entity this service describes. For example, here we say that a `User` has a `firstName` and a `lastName`, which are both `string` properties, and a date of birth (`dob`), which is a `datetime` property. Property declarations can either have a [*primative* datatype](primitives), or be a *foreign key* to another service.
```
`date` not `datetime`?

xsanda commented on TempleEight/docs/getting-started.md:
```diff
@@ -4,3 +4,400 @@ title: Getting Started
+
+#### Permission Metadata
+
+Another type of metadata, present in every service are permission based metadata. These consist of `#readable`, `#writable`, and `#enumerable`. They respectfully define which other entities can read, write and enumerate the values stored in this service. 
```
> respectfully

With respect, I think you’ve chosen the wrong word

xsanda replied:
```suggestion
Another type of metadata, present in every service are permission based metadata. These consist of `#readable`, `#writable`, and `#enumerable`. They define which other entities can read, write and enumerate respectively the values stored in this service. 
```

xsanda commented on TempleEight/docs/getting-started.md:
```diff
@@ -4,3 +4,400 @@ title: Getting Started
+  // Implicit `created_by` field, which will be Auth/User ID of creator
+  userOne: User;
+  userTwo: User;
+  matchedOn: datetime @serverSet;
```
We haven’t explained `@serverSet` yet

xsanda commented on TempleEight/docs/getting-started.md:
```diff
@@ -4,3 +4,400 @@ title: Getting Started
+
+#### CreatedBy Attributes
+
+As well as implicit `id` attributes, some services also get an implicit `createdBy` attribute generated. Those services are ones that do not contain an `#auth` block, but are contained in a Templefile where some service(s) do have an `#auth` block. An example of this can be seen in the second service block of the example `DatingApp` Templefile. The `Match` service stores an implicit `createdBy` a ...
```
This doesn’t sound deterministic enough. In a project where at least one service has an auth block, every service without its own auth block will have this attribute.

xsanda commented on TempleEight/docs/getting-started.md:
````diff
@@ -4,3 +4,400 @@ title: Getting Started
+Now we can view the generated outputs of our project:
+
+```
+~/Documents/temple-tutorial ❯❯❯ tree
````
This is going to confuse people if they don’t have `tree` installed, and don’t know what it is

xsanda commented on TempleEight/docs/getting-started.md:
````diff
@@ -4,3 +4,400 @@ title: Getting Started
+```
+/auth/
+/match/
+/dating-app-user/
````
I’d be inclined to use an example that won’t be renamed, and only mention renaming as a side note at the end.

xsanda commented on TempleEight/docs/getting-started.md:
````diff
@@ -4,3 +4,400 @@ title: Getting Started
+```
+
+These directories contain the Go code for each microservice, one per folder. This includes the `auth` service, which was generated due to the `#auth` block in the `User` service.
+A full description of these can be seen in the [Golang Reference](golang)
````
Is this a relative link, or meant to be a Markdown footer link?

xsanda replied:
Full stops please

Lewiky commented on TempleEight/docs/getting-started.md:
```diff
@@ -4,3 +4,400 @@ title: Getting Started
+
+For a full reference of the Templefile language, check out the [Templefile Spec](templefile-spec.md).
+
+We're going to work from an example Templefile for this tutorial. It describes a very simple dating app with two microservices. Here it is:
```
> It's kinda hard without syntax highlighting to understand exactly what's going on - I know there's `prism.js` support, that might be worth looking into?

I did have a look into it, seems possible to write custom syntax highlighting but I wouldn't know where to start with regexes etc, might be one for @xsanda

xsanda commented on TempleEight/docs/getting-started.md:
````diff
@@ -4,3 +4,400 @@ title: Getting Started
+19 directories, 38 files
+```
+
+### Output Breakdown
````
I wonder if this would serve better as a table?

xsanda commented on TempleEight/docs/getting-started.md:
````diff
@@ -4,3 +4,400 @@ title: Getting Started
+This registers as a new user. The JWT returned to us here needs to be included in all future requests made.
+
+```
+~/Documents/temple-turotial ❯❯❯ curl -X POST $KONG_ENTRY/api/dating-app-user -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE1ODY4NzI3MzQsImlkIjoiZTk3NWRkNTktN2Q4ZS0xMWVhLWFmYmUtMDI0MmFjMWYwMDAyIiwiaXNzIjoibnVxSmJZeWE5YnBWcTNDTnpYWUFTYWVPYzF5c2twOTAifQ.FvigKn3xf-EzQ5qSDdDdF_bySwA8o_uBQrdkkNOZauE' -d '{"firstName": "John", "lastName": "Smith", "dob": "1996-12-19T16:39:57- ...
````
turotial

xsanda commented on TempleEight/docs/getting-started.md:
````diff
@@ -4,3 +4,400 @@ title: Getting Started
+Another tool contained in the Temple CLI is the automated testing tool. Temple will take the data model defined in the Templefile, as well as the generated sources and will automatically make requests that meet the format of the data, verifying that the returned values is as expected. It can be a very useful way of end-to-end testing the application, verifying all components are functioning corre ...
+
+```
+~/Documents/temple-turotial ❯❯❯ temple test dating-app.temple
````
turotial

xsanda commented on TempleEight/docs/getting-started.md:
```diff
@@ -4,3 +4,400 @@ title: Getting Started
+    ✅ Auth register
+    ✅ Auth login
+🧪 Testing DatingAppUser service
+    ✅ DatingAppUser create
```
Side note: how bad do these emojis look on windows/linux?

xsanda commented:
>     * would it be possible to have 1 sentence per line?
>       I think it'd make it much easier to review - open to opinions on that though.

❤️

Lewiky commented:
Ok so I've rewritten the getting-started guide to be a much smaller, minimal example. For now, I've also added the installation guide and some other bits and pieces to get all the links working. I've added the stub files for the later sections of the tutorial that i'll write next, again so all the links are working and you can browse the docs at their full rn

jaylees14 commented on TempleEight/docs/guide/getting-started.md:
````diff
@@ -0,0 +1,281 @@
+In order to make a request to the application, we send a HTTP request to the Kong entrypoint URL with the correct data. 
+
+```
+~/Documents/temple-tutorial ❯❯❯ curl -X POST  $KONG_ENTRY/api/example-service -d '{"foo" : "Hello Temple!", "bar" : 123}'
````
Also worth explaining where this `/api/` came from?

xsanda commented on TempleEight/docs/guide/getting-started.md:
````diff
@@ -0,0 +1,281 @@
+
+```
+~/Documents/temple-tutorial ❯❯❯ temple generate example.temple
+What should the Go module name be? (expected format "github.com/username/repo")
````
I wonder if this could also be specified within the Templefile?

Lewiky replied:
I think it would be have to be an argument to the #language block or something, since we only need it specifically when go is the chosen lang

xsanda commented on TempleEight/docs/guide/getting-started.md:
````diff
@@ -0,0 +1,281 @@
+Now we can view the generated outputs of our project:
+
+```
+~/Documents/temple-tutorial ❯❯❯ ls
````
or at least `ls -1` for newlines

xsanda commented on TempleEight/docs/guide/getting-started.md:
````diff
@@ -0,0 +1,281 @@
+/deploy.sh
+```
+
+This shell script is an automated way to deploy the application **for local development** , including running all initialization steps and setting environment variables. 
````
```suggestion
This shell script is an automated way to deploy the application **for local development**, including running all initialization steps and setting environment variables. 
```

xsanda commented on TempleEight/docs/guide/getting-started.md:
```diff
@@ -0,0 +1,281 @@
+
+Now that we've generated some application code, we can run it and test that it works as we intended. 
+
+Make sure the Docker daemon is running and follow along.
```
```suggestion
Make sure the Docker daemon is running ([installation instructions](https://docs.docker.com/get-docker/)) and follow along.
```

xsanda commented on TempleEight/docs/guide/getting-started.md:
````diff
@@ -0,0 +1,281 @@
+localhost:8000
+```
+
+This address is the Kong entrypoint. By addressing all of our API requests here, Kong will automatically forward it on to the correct microservice.
````
Anything we can steal from the poster?

xsanda commented on TempleEight/docs/installation.md:
```diff
@@ -0,0 +1,55 @@
+
+## Installation Instructions
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
```
Importing works mid-markdown? What is this madness?

Lewiky replied:
mdx bb

xsanda commented on TempleEight/docs/guide/getting-started.md:
```diff
@@ -0,0 +1,281 @@
+
+Property declarations tell us about the kind of data belonging to the entity this service describes. 
+For example, here we say that `ExampleService` has a `foo`  which is a `string` property, and a `bar`, which is an `int` property. 
+Property declarations can either be a [*primitive* datatype](primitives), or be a [*foreign key*](cross-service-coms) to another service.
```
Do these links work? Aren’t they in different folders?

Lewiky replied:
Yup they work, Docusaurus does some magic with resolving the document ids

xsanda commented on TempleEight/docs/guide/getting-started.md:
```diff
@@ -0,0 +1,281 @@
+
+Property declarations tell us about the kind of data belonging to the entity this service describes. 
+For example, here we say that `ExampleService` has a `foo`  which is a `string` property, and a `bar`, which is an `int` property. 
+Property declarations can either be a [*primitive* datatype](primitives), or be a [*foreign key*](cross-service-coms) to another service.
```
Do these links work? Aren’t they in different folders?

Lewiky commented on TempleEight/docs/guide/getting-started.md:
```diff
@@ -0,0 +1,281 @@
+
+## Running the application
+
+Now that we've generated some application code, we can run it and test that it works as we intended. 
```
Lol good point

jaylees14 commented on TempleEight/docs/guide/getting-started.md:
````diff
@@ -0,0 +1,281 @@
+Let's check that the services are running:
+
+```
+~/Documents/temple-tutorial ❯❯❯ docker ps
````
@Lewiky How about  `docker ps --format '{{.Image}}'`

Lewiky replied:
Yeah that works!
